src/main/java/org/lsc/jndi/JndiServices.java

```diff
@@ -1180,48 +1159,74 @@ public Map<String, LscDatasets> doGetAttrsList(final String base,
-					ctx.setRequestControls(new Control[]{
-										new PagedResultsControl(pageSize, pagedResultsResponse, Control.CRITICAL)});
-				}
 
+				pagedResultsResponse = pagination();
 			} while (pagedResultsResponse != null);
-
-			// clear requestControls for future use of the JNDI context
-			if (requestPagedResults) {
-				ctx.setRequestControls(null);
-			}
+			contextRequestControls();
         } catch (CommunicationException e) {
             // Avoid handling the communication exception as a generic one
             throw e;
         } catch (ServiceUnavailableException e) {
             // Avoid handling the service unavailable exception as a generic one
             throw e;
-		} catch (NamingException e) {
+		} catch (NamingException | IOException e) {
 			// clear requestControls for future use of the JNDI context
-			ctx.setRequestControls(null);
-			LOGGER.error(e.toString());
-			LOGGER.debug(e.toString(), e);
-			
-		} catch (IOException e) {
-			// clear requestControls for future use of the JNDI context
-			ctx.setRequestControls(null);
+			contextRequestControls();
 			LOGGER.error(e.toString());
 			LOGGER.debug(e.toString(), e);
 		}
 		return res;
 	}
 
+	/**
+	 * Obtain pagination cookie to retrieve all elements in search request.
+	 * @return paging cookie
+	 * @throws IOException
+	 * @throws NamingException
+	 */
+	public byte[] pagination() throws IOException, NamingException {
+		byte[] pagedResultsResponse = null;
+		Control[] respCtls = ctx.getResponseControls();
+		if (respCtls != null) {
+			for(Control respCtl : respCtls) {
+				if (respCtl instanceof PagedResultsResponseControl) {
+					pagedResultsResponse = ((PagedResultsResponseControl) respCtl).getCookie();
+				}
+			}
+		}
+		if (pagedResultsResponse != null) {
+			ctx.setRequestControls(new Control[]{
+					new PagedResultsControl(pageSize, pagedResultsResponse, Control.CRITICAL)});
+		}
+		return pagedResultsResponse;
+	}
+
+	/**
+	 * Applying request controls such as pageSize and sortedBy for LDAP Context.
+	 */
+	public void contextRequestControls() {
+		try {
+			// Setting global pageSize variable
+			String pageSizeStr = (String) ctx.getEnvironment().get("java.naming.ldap.pageSize");
+			if (pageSizeStr != null && Integer.parseInt(pageSizeStr) > -1) {
+				pageSize = Integer.parseInt(pageSizeStr);
+				List<PagedResultsControl> requestControls = new ArrayList<>();
+				requestControls.add(new PagedResultsControl(pageSize, Control.CRITICAL));
+				ctx.setRequestControls(requestControls.toArray(new Control[requestControls.size()]));
+			}
+
+			// Setting global sortedBy variable
+			sortedBy = (String) ctx.getEnvironment().get("java.naming.ldap.sortedBy");
+			if (sortedBy != null) {
+				List<SortControl> requestControls = new ArrayList<>();
+				requestControls.add(new SortControl(sortedBy, Control.CRITICAL));
+				ctx.setRequestControls(requestControls.toArray(new Control[requestControls.size()]));
```

Comment:
Here, I think you should prepare a global requestControls list, and call setRequestControls only once.

Reply:
Fixed